vllm/v1/offloading/worker/cpu_gpu.py

```diff
@@ -0,0 +1,141 @@
+                 gpu_block_size: int, cpu_block_size: int, num_cpu_blocks: int,
+                 gpu_caches: list[torch.Tensor]):
+        assert cpu_block_size % gpu_block_size == 0
+        self.block_size_factor = cpu_block_size // gpu_block_size
+
+        self.attn_backend = attn_backend
+        self.gpu_caches: list[torch.Tensor] = gpu_caches
+
+        # cuda streams for gpu->cpu and cpu->gpu
+        self.d2h_stream = torch.cuda.Stream()
+        self.h2d_stream = torch.cuda.Stream()
+
+        # job_id -> transfer cuda event
+        self.transfer_events: dict[int, torch.cuda.Event] = {}
+        # list of cuda events available for re-use
+        self.events_pool: list[torch.cuda.Event] = []
+
+        pin_memory = is_pin_memory_available()
+
+        # allocate cpu tensors
+        logger.info("Allocating %d CPU tensors...", len(gpu_caches))
+        self.cpu_caches: list[torch.Tensor] = []
+        for gpu_tensor in gpu_caches:
+            gpu_shape = gpu_tensor.shape
+            assert len(gpu_shape) >= 4  # (2, num_blocks, ..., ...)
+            assert gpu_shape[0] == 2
+
+            cpu_shape = list(gpu_shape)
+            cpu_shape[1] = num_cpu_blocks * self.block_size_factor
+
+            logger.debug("Allocating CPU tensor of shape %r", cpu_shape)
+            self.cpu_caches.append(
+                torch.zeros(cpu_shape,
+                            dtype=gpu_tensor.dtype,
+                            device="cpu",
+                            pin_memory=pin_memory))
+
+    def transfer_async(self, job_id: int, spec: TransferSpec) -> bool:
+        src_spec, dst_spec = spec
+        if isinstance(src_spec, CPULoadStoreSpec):
+            assert isinstance(dst_spec, GPULoadStoreSpec)
+            stream = self.h2d_stream
+            src_caches = self.cpu_caches
+            dst_caches = self.gpu_caches
+            src_block_size_factor = self.block_size_factor
+            dst_block_size_factor = 1
+        else:
+            assert isinstance(src_spec, GPULoadStoreSpec)
+            assert isinstance(dst_spec, CPULoadStoreSpec)
+            stream = self.d2h_stream
+            src_caches = self.gpu_caches
+            dst_caches = self.cpu_caches
+            src_block_size_factor = 1
+            dst_block_size_factor = self.block_size_factor
+
+        src_blocks = src_spec.block_ids
+        dst_blocks = dst_spec.block_ids
+        assert src_blocks.ndim == 1
+        assert dst_blocks.ndim == 1
+
+        dst_sub_blocks_to_skip = (-src_blocks.size(0) % dst_block_size_factor)
+
+        assert (src_blocks.size(0) *
+                src_block_size_factor == dst_blocks.size(0) *
+                dst_block_size_factor - dst_sub_blocks_to_skip)
+
+        src_to_dst = torch.column_stack(
+            (expand_block_ids(src_blocks, src_block_size_factor),
+             expand_block_ids(dst_blocks,
+                              dst_block_size_factor)[dst_sub_blocks_to_skip:]))
```

Comment:
This looks good!

We could potentially optimize further by creating the empty final tensor first and having the `expand_block_ids` methods write into that rather than each creating an intermediate concatenated tensor.

Reply:
I implemented something like this now.